spanner/batchwriter_test.go

```diff
@@ -0,0 +1,234 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spanner
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	sp "cloud.google.com/go/spanner"
+	"github.com/stretchr/testify/assert"
+)
+
+// TestFlush tests NewBatchWriter, AddRow and Flush.
+func TestFlush(t *testing.T) {
+	tests := []struct {
+		count       int
+		rowSize     int
+		writeLimit  int64
+		name        string
+		badRowIndex map[int]bool // Identifies which rows are bad (by index).
+	}{
+		{count: 1, rowSize: 5, writeLimit: 40, name: "One write"},
+		{count: 50000, rowSize: 5, writeLimit: 40, name: "Many writes"},      // Forces split based on mutation count.
+		{count: 100, rowSize: 1 << 20, writeLimit: 40, name: "Large writes"}, // Forces split based on byte size.
+		{count: 50000, rowSize: 5, writeLimit: 5, name: "Write limit"},       // Forces write-limiting.
+		{count: 50, rowSize: 5, writeLimit: 40, name: "Bad rows", badRowIndex: map[int]bool{6: true, 17: true}},
+	}
+	config := BatchWriterConfig{
+		BytesLimit: 100 << 20,
+		Verbose:    false,
+		RetryLimit: 1000,
+	}
+	for _, tc := range tests {
+		data, limit := generateRows(tc.count, tc.rowSize)
+		goodRows, badRows := partitionRows(tc.badRowIndex, data)
+		badMutations := toMutations(badRows)
+		var mutex = &sync.Mutex{}
```

Comment:
Can this not be ``mutex := &sync.Mutex{}``?

Reply:
Done x 2.
Thanks!